### ISSUE TITLE
gradients are inconsistently vectorized

### ISSUE THREAD

**Issue by jkitchin** [google/tangent#30]
Sometimes gradients are vectorized, and sometimes they are not. Consider this example where the gradient is vectorized (ie. array in, array of derivatives out).

```
def f(x):
    return x**2

df = tangent.grad(f)

print(df(np.array([0, 1, 2])))
```
This comes out like I would expect.

:RESULTS:
[ 0.  2.  4.]
:END:

Compare it to this:

```
def f1(x):
    return x + 2.0 * np.cos(x)
# df/dx = 1 - 2*sin(x)

df1 = tangent.grad(f1)

x = np.array([0.0, 1.0, 2.0])
print(df1(x)) # It is not clear this is even correct.
print(1 - 2 * np.sin(x))

# A vectorized version
df1v = np.vectorize(tangent.grad(f1))
print(df1v(np.array([0, 1, 2])))
```

:RESULTS:
-2.50153682327
[ 1.         -0.68294197 -0.81859485]
[ 1.         -0.68294197 -0.81859485]

:END:

It is not clear that df1 even returns the right answer for the array.

This seems important because an obvious thing one might want to do is pass the tangent.grad function to scipy.optimize.fsolve. But fsolve requires the fprime function to take array arguments, and return an array of derivatives.

**Comment by mdanatg** [comment]
This looks like a case where we infer the return shape incorrectly.

For a temporary workaround, this should ensure the result has the proper shape:

    df1(x, b_return=np.ones_like(x))

In general, you may write `df(<arguments>, ones(<shape of f's return value>))`.

**Comment by alexbw** [comment]
I think this is a misnomer on our part -- `tangent.grad` in this case is not computing a gradient at all, but a general vector-Jacobian product. Unless your original function returns a scalar value, as @mdanatg states, you will have to supply the seed derivative yourself. For gradients, this is just the scalar `1`, so we often forget about it, and in fact, provide a kwarg default for it.  

Perhaps in the future, `tangent.grad` will check for scalar-valued output, and error out otherwise. We could then expose a more general `tangent.vjp` and `tangent.jvp` (Jacobian-vector product, which is what forward-mode autodiff computes), that does not have a default argument for the seed vector.

**Comment by alexbw** [comment]
More concretely, if you wanted to compute individual gradients for each of these dimensions, you should probably use `np.eye` as the seed vector, if you want to access each row of the Jacobian separately. Otherwise, if you want the sum of the rows of the Jacobian (which we often want in batch-mode training of neural nets), you can either provide `np.ones_like` as stated above, or add some kind of a reduction at the end of your function (adding np.sum as a final computation would have the same effect as `np.ones_like`, but you might want another reduction).

**Comment by jkitchin** [comment]
So, if I have a function:  f(x_array)

then:
df = tangent.grad(f)

can be called as:
df(x_array, b_return=np.ones_like(x_array))

and it will effectively return the Jacobian of f evaluated at each point in x_array?

but 

df(x_array) will return some kind of sum of the Jacobian matrix?

**Comment by alexbw** [comment]
Hey @jkitchin, I'm looking more into this. It looks like Tangent is improperly handling jacobian-matrix and matrix-jacobian products (it looks like the vector version is doing fine, though). Rows of the Jacobian are somewhere being summed before the final output is returned.

If your function `f` is e.g. `def f(x): return x*x`, then `df(np.array([1,2,3]), np.eye(3))` should return the full 3x3 Jacobian, which should be zeros, except for `[2,4,6]` on the diagonal. That's not happening. However, if you loop `[df(np.array([1,2,3]), np.eye(3)[i]) for i in range(3)]`, that does do the right thing, and gives you the Jacobian, one row at a time.

**Comment by jkitchin** [comment]
Here is what I see:

```
def f(x):
    return x**2

df = tangent.grad(f)
```

I would have thought this would give 12.

`print(df(np.array([1, 2, 3]), 1))`
[2 4 6]

These give what I would expect.

`print(df(np.array([1, 2, 3]), np.ones(3)))`
[ 2.  4.  6.]


`print(df(np.array([1, 2, 3]), np.eye(3)))`
[[ 2.  0.  0.]
 [ 0.  4.  0.]
 [ 0.  0.  6.]]

In the original example at the top, the -2.5... result seems incorrect if it should be a sum of the three numbers. e.g. the derivatives at [0, 1, 2] are [ 1.         -0.68294197 -0.81859485]. I don't know if that is related, or something I still don't understand.

**Comment by alexbw** [comment]
Why would you expect the first example to return 12? Reverse-mode should only be able to sum over rows of the Jacobian, not columns. 

If we run `tangent.grad(f,verbose=1)`, you can see the source code.

```python
def dfdx(x, bx_to_the_2=1.0):
    # Grad of: x_to_the_2 = x ** 2
    _bx = 2 * x * bx_to_the_2
    bx = _bx
    return bx
```

**Comment by alexbw** [comment]
We discussed a little bit, and we're going to update the API to avoid this confusion. `df = tangent.grad(f)` should technically only be valid for gradients, where some function `f` produces scalar output. We're going to assert that the output of `f` is scalar when using `tangent.grad`. The more general vector-Jacobian product should work for output values of any size, but the seed derivative should match the output size.  This will be exposed in a new function `tangent.vjp` (and an accompanying forward-mode function `tangent.jvp`)

**Comment by jkitchin** [comment]
I thought it might be 12 in part from the confusion of what grad is doing, and the relationship of the output to the shape of b_return. It looks like a misunderstanding on my part.
